src/main.rs

```diff
@@ -95,6 +102,58 @@ fn publish(request: Json<PublishRequest>, auth: TokenAuth) -> ApiResult<EmptyRes
     Ok(Json(EmptyResponse))
 }
 
+#[post(
+    "/upload_project?<forc_version>",
+    format = "application/gzip",
+    data = "<tarball>"
+)]
+async fn upload_project(
+    db: &State<Database>,
+    pinata_client: &State<PinataClientImpl>,
+    forc_version: &str,
+    mut tarball: TempFile<'_>,
+) -> ApiResult<UploadResponse> {
+    // Install the forc version if it's not already installed.
+    let forc_path_str = format!("forc-{forc_version}");
+    let forc_path = PathBuf::from(&forc_path_str);
+    fs::create_dir_all(forc_path.clone()).unwrap();
+    let forc_path = fs::canonicalize(forc_path.clone()).unwrap();
+
+    install_forc_at_path(forc_version, &forc_path)?;
+
+    // Create an upload ID and temporary directory.
+    let upload_id = Uuid::new_v4();
+    let upload_dir_str = format!("tmp/uploads/{}", upload_id);
+    let upload_dir = Path::new(&upload_dir_str);
+
+    fs::create_dir_all(upload_dir).unwrap();
+
+    // Persist the file to disk.
+    let orig_tarball_path = upload_dir.join("original.tgz");
+    tarball
+        .persist_to(&orig_tarball_path)
+        .await
+        .map_err(|_| ApiError::Upload(UploadError::SaveFile))?;
+
+    // Handle the project upload and store the metadata in the database.
+    let upload = handle_project_upload(
+        upload_dir,
+        &upload_id,
+        &orig_tarball_path,
+        &forc_path,
+        forc_version.to_string(),
+        pinata_client.inner(),
+    )
+    .await?;
+
+    let _ = db.conn().insert_upload(&upload)?;
+
+    // Clean up the temporary directory.
+    fs::remove_dir_all(upload_dir).unwrap();
```

Comment:
Can we introduce an error type that we can use instead of unwrapping on the filesystem calls?

Reply:
or change these to an expect?

Reply:
done